mmdet/datasets/pipelines/auto_augment.py

```diff
@@ -65,8 +68,251 @@ def __init__(self, policies):
+        """
+        return cv2.warpAffine(
+            data,
+            trans_matrix,
+            dsize=out_size,  # dsize takes input size as order (w,h).
+            flags=flags,
+            borderMode=borderMode,
+            borderValue=fill_val)
+
+    def __call__(self, results, min_size=0.0, neg_offset_prob=0.5):
+        """Call function to translate images, bounding boxes, masks and
+        semantic segmentation maps.
+
+        Args:
+            results (dict): Result dict from loading pipeline.
+            min_size (int | float): Minimum pixel size for the
+             translated bboxes.
+            neg_offset_prob (float): The probability that turns the
+             offset negative.
+
+        Returns:
+            dict: Translated results.
+        """
+        if np.random.rand() > self.prob:
+            return results
+
+        offset = -self.offset if np.random.rand(
+        ) < neg_offset_prob else self.offset
+        # the transformation matrix of cv2
+        if self.axis == 'x':
+            trans_matrix = np.float32([[1, 0, offset], [0, 1, 0]])
+        else:
+            trans_matrix = np.float32([[1, 0, 0], [0, 1, offset]])
+
+        self._translate_img(results, trans_matrix, fill_val=self.fill_val)
+        self._translate_bboxes(results, offset)
+        # fill_val set to 0 for background of mask.
+        self._translate_masks(results, trans_matrix, fill_val=0)
+        # fill_val set to ``seg_ignore_label`` for the ignored value
+        # of segmentation map.
+        self._translate_seg(
+            results, trans_matrix, fill_val=self.seg_ignore_label)
+        self._filter_invalid(results, min_size=min_size)
+        return results
+
+    def _translate_img(self, results, trans_matrix, fill_val):
+        """Translate images horizontally or vertically, according to
+        ``trans_matrix``."""
+        for key in results.get('img_fields', ['img']):
+            results[key] = self.warpAffine(results[key], trans_matrix,
+                                           results[key].shape[:2][::-1],
+                                           fill_val)
+
+    def _translate_bboxes(self, results, offset):
+        """Shift bboxes horizontally or vertically, according to ``offset``."""
+        h, w, c = results['img_shape']
+        for key in results.get('bbox_fields', []):
+            min_x, min_y, max_x, max_y = np.split(
+                results[key], results[key].shape[-1], axis=-1)
+            if self.axis == 'x':
+                min_x = np.maximum(0, min_x + offset)
+                max_x = np.minimum(w, max_x + offset)
+            else:
+                min_y = np.maximum(0, min_y + offset)
+                max_y = np.minimum(h, max_y + offset)
+
+            # clip box
+            min_x = np.clip(min_x, 0, w)
+            min_y = np.clip(min_y, 0, h)
+            max_x = np.clip(max_x, 0, w)
+            max_y = np.clip(max_y, 0, h)
```

Comment:
It seems `clip bbox` is redundant? Since already use `np.maximum` and `np.minimum`

Reply:
Yeah, clip bbox here is redundant.